src/core/ddsi/tests/locators.c

```diff
@@ -110,12 +110,15 @@ CU_Theory ((enum transport_selector tr), ddsi_locator_from_string, ipv4_invalid)
   snprintf (astr, sizeof (astr), "%s/:1234", fact->m_typename);
   res = ddsi_locator_from_string (&gv, &loc, astr, fact);
   CU_ASSERT_FATAL (res == AFSR_INVALID || res == AFSR_UNKNOWN);
-  snprintf (astr, sizeof (astr), "%s/[1.2.3.4]", fact->m_typename);
-  res = ddsi_locator_from_string (&gv, &loc, astr, fact);
-  CU_ASSERT_FATAL (res == AFSR_INVALID || res == AFSR_UNKNOWN);
-  snprintf (astr, sizeof (astr), "%s/[1.2.3.4]:1234", fact->m_typename);
-  res = ddsi_locator_from_string (&gv, &loc, astr, fact);
-  CU_ASSERT_FATAL (res == AFSR_INVALID || res == AFSR_UNKNOWN);
+  // The following tests are disabled because they run into a timeout
+  // when calling getaddrinfo, and this timeout depends on the configured
+  // timeout and retries in the OS resolve configuration
+  // snprintf (astr, sizeof (astr), "%s/[1.2.3.4]", fact->m_typename);
+  // res = ddsi_locator_from_string (&gv, &loc, astr, fact);
+  // CU_ASSERT_FATAL (res == AFSR_INVALID || res == AFSR_UNKNOWN);
+  // snprintf (astr, sizeof (astr), "%s/[1.2.3.4]:1234", fact->m_typename);
+  // res = ddsi_locator_from_string (&gv, &loc, astr, fact);
+  // CU_ASSERT_FATAL (res == AFSR_INVALID || res == AFSR_UNKNOWN);
```

Comment:
Perhaps it's better to use example IP addresses for tests and instead of commenting these move them into a separate test and use `.disabled = true`? Don't know if that makes sense...

Reply:
What do you mean with 'example IP addresses'? The IP addresses in the lines I commented out are handled as a host name (because of the [ and ]) and looked up via DNS (`ddsrt_gethostbyname`), which obviously fails (and causes the timeout). So I think it doesn't really matter what IP address is used here. 

On disabling the tests: if we plan to fix this in future, I agree that it would make sense to move these tests lines to a separate test and disable that test. But I thinks that not the case here (as there is no reasonable fix because it depends on OS settings), and I don't really like having more tests that report 'disabled' in the test output (which makes it harder to scan through the test output to find failures).

Reply:
It would be better if the tests use IP addresses ([RFC5737](https://tools.ietf.org/html/rfc5737)) reserved for documentation purposes. Others might be routed. Not that it makes much difference here, it's just good practice. Since you didn't introduce them, I don't require you to change them for this PR. Just thought might as well change it while you're at it(?)

If you move the tests that take a bit longer into a separate test and disable them in Travis, that'd be better. The `disabled` status is not ideal for reading, but I think that's only because the overview is not ideal to start with(?) The fact that these tests will be run locally if not explicitly disabled is added value imho, if not, we should remove them entirely. Either way, commenting them out doesn't seem the best way to go.

Reply:
Ok, I'll replace the 1.2.3.4 addresses (and those which are used as host name with [ ]) with 192.0.2.0. 

About disabling the tests: it's just the other way around: on Travis these tests run fine, because apparently the timeout for DNS resolving is a small value (or DNS resolving is disabled on Travis builds?). On my Ubuntu machine with default settings it gives a timeout because of the name resolving timeout settings.

Reply:
Since we're trying to mimic `gethostbyname` we could probably get away with checking if there are any non-hostname characters in there and simply refuse to resolve if there are, but I'll have to look into that a bit. I'll put it on my list of things todo. Since I'll have to look into OpenSSL a bit to determine if the Coverity fixes are correct, I'll do it later today or tomorrow.

Reply:
What about the following change to `ddsi_ipaddr.c`? Would this be acceptable @eboasson? Should take care of one of the time outs?

```diff --git a/src/core/ddsi/src/ddsi_ipaddr.c b/src/core/ddsi/src/ddsi_ipaddr.c
index 4aa408bd..2d731e0b 100644
--- a/src/core/ddsi/src/ddsi_ipaddr.c
+++ b/src/core/ddsi/src/ddsi_ipaddr.c
@@ -138,26 +138,25 @@ enum ddsi_locator_from_string_result ddsi_ipaddr_from_string (const struct ddsi_
     }
   }
 
-#if DDSRT_HAVE_IPV6
-  if (af == AF_INET6)
-  {
-    if (copy[0] == '[') {
-      // strip brackets: last character before the port must be a ']',
-      // in the absence of a port, the last character in the string.
-      ipstr = copy + 1;
-      if (portstr == NULL) {
-        if (copy[len - 1] != ']')
-          return AFSR_INVALID;
-        copy[len - 1] = 0;
-      } else {
-        assert (portstr > copy);
-        if (portstr[-1] != ']')
-          return AFSR_INVALID;
-        portstr[-1] = 0;
-      }
+  if (copy[0] == '[') {
+    // strip brackets: last character before the port must be a ']',
+    // in the absence of a port, the last character in the string.
+    ipstr = copy + 1;
+    if (portstr == NULL) {
+      if (copy[len - 1] != ']')
+        return AFSR_INVALID;
+      copy[len - 1] = 0;
+    } else {
+      assert (portstr > copy);
+      if (portstr[-1] != ']')
+        return AFSR_INVALID;
+      portstr[-1] = 0;
     }
-  }
+#if DDSRT_HAVE_IPV6
+    if (af != AF_INET6)
 #endif
+      return AFSR_INVALID;
+  }
 
   if (ddsrt_sockaddrfromstr(af, ipstr, (struct sockaddr *) &tmpaddr) != 0) {
 #if DDSRT_HAVE_DNS

```

Reply:
I suppose that would work, but I do wonder if it wouldn't make more sense to allow the bracketed notation also for IPv4. It's not like those brackets would cause any problems.

I am also generally in favour of rejecting malformed inputs early. I imagine that it would be easy enough to check whether the the string could possibly be a hostname as you suggested, and one would think that'd be the nicest solution for this particular problem. I suppose you know the rules for hostnames off the top of your head, so I'm not going to bother looking them up right now 😁

Reply:
So, I thought about this a bit more... Basically it's a bug in some old(er) version of Glibc, so technically it's not our bug to fix and since we haven't had any other complaints, we could choose to leave the tests and implementation as they are. However, since we don't pass AI_IDN, checking the hostname is relatively easy, just check if it's **a)** an IPv4 address or **b)** an IPv6 address or **c)** a classic hostname containing just `[a-zA-Z0-9-]`. That should take care of the problem without breaking things for users. If it does break things, we could choose to treat strings containing UTF-8 characters as IDN names and let `getaddrinfo` worry about it (or use `libidn`, which `getaddrinfo` itself probably also uses) and validate strings containing only ASCII characters the same way. An URI parser I once did can be found [here](https://github.com/ADLINK-IST/opensplice/blob/master/src/abstraction/os/common/code/os_uri.c). It's owned by ADLINK so we can copy the necessary code and re-license or you can just do something like this:
```
for (size_t i=0; name[i]; i++)
  if (name[i] != <valid-hostname-character-or-character-that-occurs-in-ip-address>)
    return <error>
```
Probably close enough for our use-case(?)

Reply:
Thanks for looking into this. I've implemented (commit 6c835b3) your suggestion of checking for valid characters before passing the value to `getaddrinfo`, I also think that this is sufficient in our situation and we don't need the complete uri parser code for this.

Reply:
@k0ekk0ek, in commit d6206c9 I've replaced `isalnum` in the host name check as you suggested, because of its locale dependent behavior